packages/@aws-cdk/aws-iotevents/lib/state.ts

```diff
@@ -0,0 +1,51 @@
+import { Event, getEventJson } from './event';
+import { CfnDetectorModel } from './iotevents.generated';
+
+/**
+ * Properties for defining a state of a detector
+ */
+export interface StateProps {
+  /**
+   * The name of the state
+   */
+  readonly stateName: string;
+
+  /**
+   * Specifies the actions that are performed when the state is entered and the `condition` is `TRUE`
+   *
+   * @default None
+   */
+  readonly onEnterEvents?: Event[];
+}
+
+/**
+ * Defines a state of a detector
+ */
+export class State {
+  /**
+   * The name of the state
+   */
+  public readonly stateName: string;
+
+  constructor(private readonly props: StateProps) {
+    this.stateName = props.stateName;
+  }
+
+  /**
+   * Return the state property JSON
+   */
+  public toStateJson(): CfnDetectorModel.StateProperty {
+    const { stateName, onEnterEvents } = this.props;
+    return {
+      stateName,
+      onEnter: onEnterEvents && { events: getEventJson(onEnterEvents) },
+    };
+  }
+
+  /**
+   * returns true if this state has at least one condition via events
+   */
+  public hasCondition(): boolean {
```

Comment:
I actually don't like this method. This is actually a `DetectorModel` concern, not a `State` concern, that this validation has to be performed.

Can we remove this method, and perform this validation in `DetectorModel` by calling `toStateJson()` on its `initialState` instead?

Reply:
If we implement this validation with `toStateJson()`, It will be as following I think:
```ts
const stateJson = props.initialState._toStateJson();

if (!stateJson.onEnter?.events.some(event => event.condition)) {
  throw new Error('Detector Model must have at least one Input with a condition');
}
```

But we can't get `onEnter?.events` because `onEnter` can be `cdk.IResolvable`, same for `events.some()` and `event.condition`.

```ts
export interface StateProperty {
  readonly onEnter?: CfnDetectorModel.OnEnterProperty | cdk.IResolvable;
}
```

Should `toStateJson()` return more exact type instead of `CfnDetectorModel.StateProperty`?

Following is the type `CfnDetectorModel.StateProperty` that `toStateJson()` return:
```ts
namespace CfnDetectorModel {
  export interface StateProperty {
    readonly onEnter?: CfnDetectorModel.OnEnterProperty | cdk.IResolvable;
    readonly onExit?: CfnDetectorModel.OnExitProperty | cdk.IResolvable;
    readonly onInput?: CfnDetectorModel.OnInputProperty | cdk.IResolvable;
    readonly stateName: string;
  }
}
```

Reply:
OK, I see the problem. I think we could just check whether `onEnter` is a `CfnDetectorModel.OnEnterProperty`, and skip the validation if it's an `IResolvable`. But I don't want you to write all of that code if a simpler alternative is possible here.

So, let's keep `_toStateJson()` as it is now, but let's make this method `@internal`, and rename it to `_onEnterEventsHaveAtLeastOneCondition()`.